src/day10.rs

```diff
@@ -1,10 +1,235 @@
+        .x_label_area_size::<i32>(0)
+        .y_label_area_size::<i32>(0)
+        .build_cartesian_2d(x_min..x_max, y_min..y_max)?; // TODO: Fix these limits automagically
+    scatter_ctx
+        .configure_mesh()
+        .disable_x_mesh()
+        .disable_y_mesh()
+        .draw()?;
+    scatter_ctx.draw_series(
+        particles
+            .iter()
+            .map(|particle| Circle::<(i32, i32), i32>::new((particle.pos.x, particle.pos.y), 5, GREEN.filled())),
+    )?;
+
+    scatter_ctx
+        .configure_series_labels()
+        .background_style(&WHITE.mix(0.8))
+        .border_style(&BLACK)
+        .draw()?;
+
+    Ok(())
+}
+
+fn find_average_pos(particles: &Vec<Particle>) -> Coord {
+    let mut coord = Coord {x:0, y:0};
+    for particle in particles {
+        coord.x += particle.pos.x;
+        coord.y += particle.pos.y;
+    }
+    coord / Coord::new_scale(particles.len() as i32)
+}
+
+fn in_range_of_target(particles: &mut Vec<Particle>, target: &Coord, range:i32) -> bool {
+    let result = particles.iter().all(|particle| particle.pos.distance(target) < range);
+    result
+}
+
+fn lineyness(positions: &HashSet<Coord>) -> f32 {
+    let mut count = 0;
+    for pos in positions.clone().into_iter() {
+        if positions.contains(&pos.left()) && positions.contains(&pos.right()) {
+            count += 1;
+        }
+        if positions.contains(&pos.above()) && positions.contains(&pos.below()) {
+            count += 1;
+        }
+    }
+    
+    (count as f32) / (positions.len() as f32)
+}
+
+fn in_final_position(particles: &Vec<Particle>) -> bool {
+
+    // find number of overlapping particles
+    let mut positions:HashSet<Coord> = HashSet::new();
+    let mut overlap_count = 0;
+    for particle in particles {
+        if !positions.insert(particle.pos) {
+            overlap_count += 1;
+        }
+    }
+
+    if overlap_count >= 5 { // overlap count cheated a little here by using debug output to estimate this value
+        if lineyness(&positions) > 0.6 { // but 60% was a genuine guess based on the problem (I did get lucky though XD)
+            return true
+        }
+    }
+    false
+}
+
+fn find_correct_positions(particles: &mut Vec<Particle>, _part_one: &mut String, part_two: &mut i32) {
+    let mut target = find_average_pos(particles);
+    while !in_range_of_target(particles, &target, 10000) {
+        particles.iter_mut().for_each(|particle| particle.move_thousand());
+        target = find_average_pos(particles);
+        *part_two += 1000;
+    }
+    while !in_range_of_target(particles, &target, 1000) {
+        particles.iter_mut().for_each(|particle| particle.move_hundred());
+        target = find_average_pos(particles);
+        *part_two += 100;
+    }
+    while !in_final_position(particles) {
+        particles.iter_mut().for_each(|particle| particle.move_one());
+        *part_two += 1;
+    }
+    particles.iter_mut().for_each(|particle| particle.pos.y *= -1);
```

Comment:
Yeah - same thing I guess (although I'm wondering if I actually avoid fewer instructions here because I don't spend time multiplying by 1 and 0, I guess it probably compiles down to be very similar).